### PR TITLE
State transitions

### PR DESCRIPTION
**Pull request by bidzyyys**
https://golemproject.atlassian.net/secure/RapidBoard.jspa?rapidView=4&modal=detail&selectedIssue=CGI-44

### CODE REVIEW

File: TranscodingDemo/Server/Tests/test_states.py [hunk 1]
```diff
@@ -23,102 +13,88 @@ def test_transition_index(self):
...
     def test_transcoding_progress(self):
-        # User sends the same request for the second time.
         self.init_test_progress()
 
         response = self.send_request("h265", "mp4", 2, "1280,720", mock_creating_task=TestRefresh.create_task_mock)
-        self.assertEqual(response.status_code, http.HTTPStatus.OK)
-        self.assertIsNone(response.context.get("ErrorMessage"))
-        
-        # Seconds request
-        response = self.send_request("h265", "mp4", 2, "1280,720", mock_creating_task=TestRefresh.create_task_mock)
```
Comment (weaselix) [inline TranscodingDemo/Server/Tests/test_states.py]: Why did you remove second request? It looks like of changing a real test.

File: TranscodingDemo/Server/applogic.py [hunk 2]
```diff
@@ -34,38 +32,39 @@ def __init__(self):
...
             session.on_start_session(request)
 
         return StateManager.create_from_json(request.session["StateManager"])
 
-    def run(self, request, new_state):
+    def validate_state_url(self, state, url):
+        if url != self.urls_map[state]:
+            raise RedirectException(self.urls_map[state])
+
+    def run(self, request):
```
Comment (weaselix) [inline TranscodingDemo/Server/applogic.py]: In the future tasks we need to get rid of that "global" method. Obviously, that PR (task) does not involve that. It is just a note.

File: TranscodingDemo/Server/applogic.py [hunk 3]
```diff
@@ -172,49 +171,54 @@ def resolution_to_array(resolution_str):
...
+        params["NumSubtasks"] = int(request_body["NumSubtasks"])
+        params["Resolution"] = Logic.resolution_to_array(request_body["Resolution"])
+        params["Name"] = uuid.uuid4().hex[:24]
+        params["OutputDir"] = settings.OUTPUT_PATH
+        params["VideoName"] = video_name
+
+        return params
+
+    @staticmethod
+    def request_task(request):
```
Comment (weaselix) [inline TranscodingDemo/Server/applogic.py]: Generally, you did changes in a very good direction. Good job. 

However, it looks like `request_task` returns `JSON`. However, user (client) expect a html response. Am I right?

File: TranscodingDemo/Server/applogic.py [hunk 4]
```diff
@@ -172,49 +171,54 @@ def resolution_to_array(resolution_str):
...
         try:
+            state = session.get_state_manager(request).state
+            if state != StateFlag.UploadingCompleted:
+                return JsonResponse(status=HTTPStatus.BAD_REQUEST, data={"ErrorMessage": "Task cannot be created in "
+                                                                                         "this state"})
+
             job = TranscodingJob.objects.select_for_update().get(id=request.session['TranscodingJob'])
-            if job.tried_to_send_to_golem:
```
Comment (weaselix) [inline TranscodingDemo/Server/applogic.py]: I do not see a reason why you removed it. We need that.

File: TranscodingDemo/Server/Tests/test_states.py [hunk 5]
```diff
@@ -130,8 +106,167 @@ def test_upload_restart_page_transcode_previous(self):
...
 
-        response = self.send_request( "h265", "mp4", 3, "1920,1080" )
-        self.assertEqual(response.status_code, http.HTTPStatus.OK)
+        response = self.send_request("h265", "mp4", 3, "1920,1080")
+        self.assertEqual(response.status_code, HTTPStatus.OK)
+
+
+class TestIndex(ClientTestCase):
+
+    def verify_index_page(self, response):
```
Comment (nieznanysprawiciel) [inline TranscodingDemo/Server/Tests/test_states.py]: Just a note for future. It's better to have each test checking only one feature to avoid writing mega tests checking everything. If test fails, it's easier to find out what doesn't work.

Here for example "test redirection from TranscodingProgress" and "test redirection from Result could be separate tests.

File: TranscodingDemo/Server/viewsmanager.py [hunk 6]
```diff
@@ -1,38 +1,31 @@
-from .states import StateFlag, StateManager
 from django.shortcuts import render
 
-
+from .states import StateFlag
 
 
 class ViewsManager:
```
Comment (nieznanysprawiciel) [inline TranscodingDemo/Server/viewsmanager.py]: It seems there is no need for this class anymore (actually it was never used :D). Maybe it would be better to just throw it out instead of refactoring it and duplicating htmls map.

File: TranscodingDemo/Server/applogic.py (hunk 3 above)
Comment (bidzyyys) [reply TranscodingDemo/Server/applogic.py]: `request_task` returns `JSON` response and it is my purpose. TranscodingProgress page is served by `GET` request. We think about making our demo more scalable - the best way to achieve it (IMO) is to serve only `JSON` responses and write another front-end app. This is the way that modern services are made.

File: TranscodingDemo/Server/Tests/test_states.py (hunk 1 above)
Comment (bidzyyys) [reply TranscodingDemo/Server/Tests/test_states.py]: With my changes, task can be requested only once by `POST` method. This request is only accepted when we are in state `ViedoReady`, in other case `BAD_REQEST` is returned. Post request changes state to `TranscodingProgress` and returns `JSON` response. If the status is `OK`, there should be another request (`GET` `/TranscodingProgress`) to get TranscodingProgress page.

File: TranscodingDemo/Server/applogic.py (hunk 2 above)
Comment (bidzyyys) [reply TranscodingDemo/Server/applogic.py]: Ok

File: TranscodingDemo/Server/applogic.py (hunk 4 above)
Comment (bidzyyys) [reply TranscodingDemo/Server/applogic.py]: I rely on state. We can request task only in `UploadingCompleted` state, requesting task in any other state is blocked and even tested :D. We can request task only once, `POST` method is responsible only for requesting, not serving `TranscodingProgress` page.

File: TranscodingDemo/Server/Tests/test_states.py (hunk 5 above)
Comment (bidzyyys) [reply TranscodingDemo/Server/Tests/test_states.py]: Good point :+1: I will refactor my tests.

File: TranscodingDemo/Server/applogic.py (hunk 3 above)
Comment (bidzyyys) [reply TranscodingDemo/Server/applogic.py]: Look 
```
$("form#request-transcoding").submit(function () {
                const formData = getJsonFromFormData($(this));
                $.ajax({
                    url: '/OnlineTranscoding/TranscodingProgress/',
                    type: 'POST',
                    contentType: 'application/json',
                    headers: {"X-CSRFToken": '{{ csrf_token }}'},
                    data: JSON.stringify(formData),
                    dataType: 'json',
                    success: function (data) {
                        window.location.replace('/OnlineTranscoding/TranscodingProgress/');
                    },
                    error: function (xhr, status, error) {
                         data = JSON.parse(xhr.responseText);
                         console.log(data.ErrorMessage);
                    },
                });
            });

```

File: TranscodingDemo/Server/viewsmanager.py (hunk 6 above)
Comment (bidzyyys) [reply TranscodingDemo/Server/viewsmanager.py]: We will throw it while making our demo stateless. Now we need to know which template should be chosen.

File: TranscodingDemo/Server/applogic.py (hunk 3 above)
Comment (weaselix) [reply TranscodingDemo/Server/applogic.py]: > We think about making our demo more scalable
 I don't see how returning `JSON` instead of `html` is made service more scalable.

But, generally I agree with you and it is very good direction to split backend and frontend completely. 

> responses and write another frontend app (in Angular or React). This is the way that modern services are made.
Yes, indeed.

File: TranscodingDemo/Server/applogic.py (hunk 4 above)
Comment (weaselix) [reply TranscodingDemo/Server/applogic.py]: It is possible to request two tasks (or more). Imagine two (or more) concurrent requests.  We need a shared lock. This is why I use lock on database.

Comment (bidzyyys) [reply TranscodingDemo/Server/applogic.py]: You are right.
